_drafts/2023-06-01-functional-error-handling-in-kotlin-part-2.md

```diff
@@ -0,0 +1,912 @@
+
+```kotlin
+// Kotlin SDK
+public inline fun <T> Result<T>.onSuccess(action: (value: T) -> Unit): Result<T> {
+    contract {
+        callsInPlace(action, InvocationKind.AT_MOST_ONCE)
+    }
+    if (isSuccess) action(value as T)
+    return this
+}
+
+public inline fun <T> Result<T>.onFailure(action: (exception: Throwable) -> Unit): Result<T> {
+    contract {
+        callsInPlace(action, InvocationKind.AT_MOST_ONCE)
+    }
+    exceptionOrNull()?.let { action(it) }
+    return this
+}
+```
+
+We can refactor our primary example to use the `onSuccess` and `onFailure` functions:
+
+```kotlin
+val notFoundJobId = JobId(42)
+val maybeSalary: Result<Double> = 
+    JobService(jobs, currencyConverter).getSalaryInEur(notFoundJobId)
+maybeSalary.onSuccess {
+    println("The salary of jobId $notFoundJobId is $it")
+}.onFailure {
+    when (it) {
+        is IllegalArgumentException -> println("The amount must be positive")
+        else -> println("An error occurred ${it.message}")
+    }
+}
+```
+
+Clearly, the output of the above code is the following since the `JobId` 42 is not present in our database:
+
+```text
+The amount must be positive
+```
+
+If we want to give both the lambda to apply in case of success and the lambda to apply in case of failure, we can use the `fold` function:
+
+```kotlin
+// Kotlin SDK
+public inline fun <R, T> Result<T>.fold(
+    onSuccess: (value: T) -> R,
+    onFailure: (exception: Throwable) -> R
+): R {
+    contract {
+        callsInPlace(onSuccess, InvocationKind.AT_MOST_ONCE)
+        callsInPlace(onFailure, InvocationKind.AT_MOST_ONCE)
+    }
+    return when (val exception = exceptionOrNull()) {
+        null -> onSuccess(value as T)
+        else -> onFailure(exception)
+    }
+}
+```
+
+As we can see, the `fold` function is not only used to apply side effects. It can be used to transform a `Result` into another type. In fact, many of the transformations we've seen so far are shorthands for applying the `fold` function in particular cases.
+
+The above example can be rewritten using the `fold` function as follows:
+
+```kotlin
+maybeSalary.fold({
+    println("The salary of jobId $notFoundJobId is $it")
+}, {
+    when (it) {
+        is IllegalArgumentException -> println("The amount must be positive")
+        else -> println("An error occurred ${it.message}")
+    }
+})
+```
+
+## 3. Composing `Result` Instances
+
+As we saw in the previous article, when dealing with _effects_, _monads_, _container types_ or whatever we want to call them, one crucial point is how we can compose and combine them. In the first part of this series, we implemented a function that returns the gap between the job salary given a job id and the maximum compensation for the same company. We called the function `getSalaryGapWithMax`.
```

Comment:
"or whatever we want to call them" implies that effects, monads and container types are the same thing, but they're not.

I'd suggest something like 
_As we saw in the previous article, we often deal with types (usually some containers) that are "chainable" or "monadic" (link to the simple monads article) in structure, so the crucial point is how we can compose and combine them._